pkg/controller/direct/colab/runtime_controller.go

```diff
@@ -0,0 +1,268 @@
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting ColabRuntime %q: %w", a.id, err)
+	}
+
+	a.actual = notebookRuntimePb
+	return true, nil
+}
+
+func normalizeReferences(ctx context.Context, reader client.Reader, obj *krm.ColabRuntime) error {
+	if obj.Spec.ColabRuntimeTemplateRef != nil {
+		if _, err := obj.Spec.ColabRuntimeTemplateRef.NormalizedExternal(ctx, reader, obj.GetNamespace()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *runtimeAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating ColabRuntime", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	if err := normalizeReferences(ctx, a.reader, a.desired); err != nil {
+		return fmt.Errorf("resolving references: %w", err)
+	}
+
+	desiredPb := ColabRuntimeSpec_ToProto(mapCtx, &a.desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	fmt.Printf("maqiuyu... create desiredPb: %+v\n", desiredPb)
+
+	req := &pb.AssignNotebookRuntimeRequest{
+		Parent:                  a.id.Parent().String(),
+		NotebookRuntimeTemplate: a.desired.Spec.ColabRuntimeTemplateRef.External,
+		NotebookRuntime:         desiredPb,
+		NotebookRuntimeId:       a.id.ID(),
+	}
+	op, err := a.gcpClient.AssignNotebookRuntime(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating ColabRuntime %s: %w", a.id.String(), err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("ColabRuntime %s waiting creation: %w", a.id, err)
+	}
+	log.V(2).Info("successfully created ColabRuntime in gcp", "name", a.id)
+
+	status := &krm.ColabRuntimeStatus{}
+	status.ObservedState = ColabRuntimeObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.PtrTo(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *runtimeAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating ColabRuntime", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	if err := normalizeReferences(ctx, a.reader, a.desired); err != nil {
+		return fmt.Errorf("resolving references: %w", err)
+	}
+
+	desiredPb := ColabRuntimeSpec_ToProto(mapCtx, &a.desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	desiredPb.Name = a.id.String()
+
+	paths := make(sets.Set[string])
+	var err error
+	paths, err = common.CompareProtoMessage(desiredPb, a.actual, common.BasicDiff)
+	if err != nil {
+		return err
+	}
+
+	if len(paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id)
+		return nil
```

Comment:
Update KRM status even though there is no field to update? (for acquiring a resource)